src/io/parquet/read/primitive/basic.rs

```diff
@@ -72,6 +72,35 @@ fn read_dict_buffer_optional<T, A, F>(
     }
 }
 
+fn read_dict_buffer_required<T, A, F>(
+    indices_buffer: &[u8],
+    additional: usize,
+    dict: &PrimitivePageDict<T>,
+    values: &mut MutableBuffer<A>,
+    validity: &mut MutableBitmap,
+    op: F,
+) where
+    T: NativeType,
+    A: ArrowNativeType,
+    F: Fn(T) -> A,
+{
+    let dict_values = dict.values();
+
+    // SPEC: Data page format: the bit width used to encode the entry ids stored as 1 byte (max bit width = 32),
+    // SPEC: followed by the values encoded using RLE/Bit packed described above (with the given bit width).
+    let bit_width = indices_buffer[0];
+    let indices_buffer = &indices_buffer[1..];
+
+    let indices = hybrid_rle::HybridRleDecoder::new(indices_buffer, bit_width as u32, additional);
+
+    for index in indices {
+        let value = op(dict_values[index as usize]);
+        values.push(value);
+    }
```

Comment:
```suggestion
    values.extend_from_trusted_len(indices.map(|index| op(dict_values[index as usize])));
```

Reply:
Tried this earlier, that didn't work:

```
  --> src/io/parquet/read/primitive/basic.rs:96:41
   |
96 |     values.extend_from_trusted_len_iter(indices.map(|index| op(dict_values[index as usize])));
   |                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `trusted_len::TrustedLen` is not implemented for `HybridRleDecoder<'_>`
   |
note: required because of the requirements on the impl of `trusted_len::TrustedLen` for `std::iter::Map<HybridRleDecoder<'_>, [closure@src/io/parquet/read/primitive/basic.rs:96:53: 96:92]>`
  --> src/trusted_len.rs:12:55
   |
12 | unsafe impl<B, I: TrustedLen, T: FnMut(I::Item) -> B> TrustedLen for std::iter::Map<I, T> {}
   |                                                       ^^^^^^^^^^     ^^^^^^^^^^^^^^^^^^^^
```

Reply:
`extend` compiles though

Reply:
uf, I am glad that method is not marked `unsafe` or this could result in UB. Thanks, `extend` is still better 👍